### ISSUE TITLE
OSError: exception: stack overflow (Appears during clean_network)

### ISSUE THREAD

**Issue by jobelanger** [worldbank/GOSTnets#35]
Received this error when running `clean_network` on a large graph. In my case I was using it for country-level datasets in Mali and Burkina Faso, running on a Windows 10 machine. My current workaround as suggested by @d3netxer is to skip the cleaning function altogether.[
<img width="423" alt="oserror_stack_overflow" src="https://user-images.githubusercontent.com/48232572/129410892-72f3a44c-9659-472e-a521-d440f7b378e3.PNG">
](url)

**Comment by d3netxer** [comment]
I'm doing some work on my horn of africa analysis. I believe that totally skipping the cleaning is why I had trouble getting O-D routes. Missing reflected edges still need to be added. I'm trying to run this separately:
```
G_full = gn.add_missing_reflected_edges(G)
``` 
will provide update...

**Comment by gracedoherty** [comment]
@d3netxer Any update on this issue? I'm getting the same error. I was going to run clean_network, then salt_long_lines, then proceed with the time-based OD creation. Sounds like we shouldn't be skipping the clean_network function if so?

I'm using the latest Africa osm pbf and clipping it to a buffered border of Senegal & The Gambia.

**Comment by d3netxer** [comment]
no, you don’t want to skip cleaning the network. But you can try running simplify junctions and adding missing reflected edges separately, and see if this works.

**Comment by gracedoherty** [comment]
Okay, thanks. simplify_junctions is the point where the stack overflow error prints. Other issues cropped up for me when I removed it from the .py file. I ran an abbreviated version of the function and it completed. I included: 

- add_missing_reflected_edges()
- custom_simplify() # the second iteration
- convert_to_MultiDiGraph()
- add_missing_reflected_edges()

And I removed the following: 
- simplify_junctions() # Stack overflow error.
- custom_simplify() # the first iteration
- remove_duplicate_edges # I forgot to document what line this was happening, but for either custom simplify or remove duplicate edges, the kernel stalls and never finishes (still "running" even overnight).

.
.
.
`#the cleaning network part
import os, sys, time

import importlib
import osmnx as ox

from shapely.ops import unary_union
from shapely.wkt import loads
from shapely.geometry import LineString, MultiLineString, Point

from .core import *


def clean_network(G, wpath = '', output_file_name = '', UTM = {'init': 'epsg:3857'}, WGS = {'init': 'epsg:4326'}, junctdist = 50, verbose = False):   
    """
    Topologically simplifies an input graph object by collapsing junctions and removing interstital nodes
    :param G: a graph object containing nodes and edges. Edges should have a property called 'Wkt' containing geometry objects describing the roads.
    :param wpath: the write path - a drive directory for inputs and output
    :param output_file_name: This will be the output file name with '_processed' appended
    :param UTM: The epsg code of the projection, in metres, to apply the junctdist
    :param WGS: the current crs of the graph's geometry properties. By default, assumes WGS 84 (epsg 4326)
    :param junctdist: distance within which to collapse neighboring nodes. simplifies junctions. Set to 0.1 if not simplification desired. 50m good for national (primary / secondary) networks
    :param verbose: if True, saves down intermediate stages for dissection
    """

    # ensures all streets are two-way
    a = add_missing_reflected_edges(G)
    
    # save progress
    if verbose is True: 
        save(a, 'a', wpath)
    
    # Finds and deletes interstital nodes based on node degree
    b = custom_simplify(a)
    
    # rectify geometry
    for u, v, data in b.edges(data = True):
        if type(data['Wkt']) == list:
                data['Wkt'] = unbundle_geometry(data['Wkt'])

    # save progress
    if verbose is True: 
        save(b, 'b', wpath)
    
    # For some reason CustomSimplify doesn't return a MultiDiGraph. Fix that here
    c = convert_to_MultiDiGraph(b)
    
    # rectify geometry again
    for u, v, data in c.edges(data = True):
        if type(data['Wkt']) == list:
            data['Wkt'] = unbundle_geometry(data['Wkt'])

    # Ensure all remaining edges are duplicated (two-way streets)
    c = add_missing_reflected_edges(c)
    
    # save final
    if verbose:
       save(c, '%s_processed' % output_file_name, wpath)
    
    print('Edge reduction: %s to %s (%d percent)' % (G.number_of_edges(), 
                                               c.number_of_edges(), 
                                               ((G.number_of_edges() - c.number_of_edges())/G.number_of_edges()*100)))
    return c`

**Comment by jwrap** [comment]
I'm working on the OSM network for Cabo Verde (downloaded from Geofabrik) and ran into the same issue. I managed to resolve it by modifying the simplify_junctions function as follows:

![image](https://user-images.githubusercontent.com/22310192/155878859-89a21cfe-6101-49ec-88a5-9291fd892132.png)

1. not sure why but attempting to change the crs of gdfnodes simply changed my geometries to POINT (inf inf)
2. unary_union was causing the stack overflow, so i used the geometry after passing .dissolve() instead

Further down the line I see that a function tries to call "ConvertToMultiDiGraph" instead of "convert_to_MultiDiGraph" so i fixed that too.

![image](https://user-images.githubusercontent.com/22310192/155878945-2bf89559-71a1-4a18-a332-8d2b9bbd73af.png)

Now the function seems to work well for me. Hope this helps!

**Comment by d3netxer** [comment]
Thanks @jwrap for your comments. Here are some responses:

- I would like to drill down why the attempting to change the crs didn't work for you. Do you know the which measure_crs you used? Also the projection/coordinates of you graph?
- I do recall isolating the stack_overflow() error to the unary_union code. It tends to pop-up in very large graphs. This seems like a great way to get around it is using the Geopandas [dissolve function ](https://geopandas.org/en/stable/docs/user_guide/aggregation_with_dissolve.html) you found? I didn't know about that.
- The current master branch does have the up-to-date [convert_to_MultiDiGraph function](https://github.com/worldbank/GOSTnets/blob/5fe249a956a6ef19a8c21906c66612ecc2bc46bf/GOSTnets/network_clean.py#L49). I recommend you uninstall GOSTnets, then pull the master branch and reference it in your code.

**Comment by Charlesfox1** [comment]
GOSTs in the shell… unary_union sucks for large graphs. I’m sorry I ever used it frankly. 

One solution I had was to split the big graph into say ten equal parts, unary union them… then join them at the end. 

Or push whole graph down to something like igraph so it does it in C…

Or use a PostgreSQL instance + something like fastapi to send the graph in node/ edge form into a DB, do geom transformation there, then send it back again… 

“

**Comment by d3netxer** [comment]
Interesting, you got me thinking now Charles...maybe Dask can be used. I'm reading there is a dask-geopandas, I never tried that before; but I want to try that now, I'm making a mental note.

**Comment by Charlesfox1** [comment]
Dask could work computationally, for sure. 

If you guys don’t have a PostGIS server set up on a box somewhere yet I’d recommend that as a real priority though. You could store every network ever made in a dedicated GOSTnets Schema… even connects to QGIS directly for super fast loading / rendering. Slap a FastAPI on top of that and you could do a load of heavy lifting on the database, even if you were on mission on the other side of the world…
